mlflow/utils/virtualenv.py

```diff
@@ -0,0 +1,259 @@
+import os
+import logging
+import subprocess
+import hashlib
+import shutil
+import uuid
+from pathlib import Path
+
+import yaml
+
+from mlflow.exceptions import MlflowException
+from mlflow.utils import process
+from mlflow.utils.environment import (
+    PythonEnv,
+    _PYTHON_ENV_FILE_NAME,
+    _CONDA_ENV_FILE_NAME,
+    _get_mlflow_env_name,
+)
+from mlflow.utils.requirements_utils import _get_package_name
+from mlflow.version import VERSION
+
+
+_MLFLOW_ENV_ROOT_ENV_VAR = "MLFLOW_ENV_ROOT"
+_IS_UNIX = os.name != "nt"
+
+
+_logger = logging.getLogger(__name__)
+
+
+def _hash(s):
+    return hashlib.sha1(s.encode("utf-8")).hexdigest()
+
+
+def _join_commands(*commands):
+    entry_point = ["bash", "-c"] if _IS_UNIX else ["cmd", "/c"]
+    sep = " && " if _IS_UNIX else " & "
+    return [*entry_point, sep.join(commands)]
+
+
+def _get_pip_install_mlflow():
+    mlflow_home = os.getenv("MLFLOW_HOME")
+    if mlflow_home:  # dev version
+        return "pip install -e {} 1>&2".format(mlflow_home)
+    else:
+        return "pip install mlflow=={} 1>&2".format(VERSION)
+
+
+def _get_mlflow_env_root():
+    """
+    Returns the root directory to store virtualenv environments created by MLflow.
+    """
+    return os.getenv(_MLFLOW_ENV_ROOT_ENV_VAR, str(Path.home().joinpath(".mlflow", "envs")))
+
+
+def _get_conda_dependencies(conda_yaml_path, exclude=()):
+    """
+    Extracts conda dependencies from a conda yaml file. Packages in `exclude` will be excluded
+    from the result.
+
+    :param conda_yaml_path: Conda yaml file path.
+    :param exclude: Packages to be excluded from the result.
+    """
+    with open(conda_yaml_path) as f:
+        conda_yaml = yaml.safe_load(f)
+    return [
+        d
+        for d in conda_yaml.get("dependencies", [])
+        if isinstance(d, str) and _get_package_name(d) not in exclude
+    ]
+
+
+def _is_pyenv_available():
+    """
+    Returns True if pyenv is available, otherwise False.
+    """
+    return subprocess.run(["pyenv", "--version"], capture_output=True, check=False).returncode == 0
```

Comment:
```suggestion
    from shutil import which
    return which("pyenv") is not None
```

https://stackoverflow.com/a/34177358/6943581